AEPMessaging/Tests/UnitTests/Cache+MessagingTests.swift

```diff
@@ -15,11 +15,272 @@ import XCTest
+        }
+        XCTAssertEqual(1, decodedSetParam.count)
+        let paramProp = decodedSetParam["newSurface"]?.first
+        XCTAssertEqual("inapp4", paramProp?.uniqueId)
+        XCTAssertEqual("inapp4", paramProp?.scope)
+        XCTAssertEqual(1, paramProp?.items.count)
+    }
+    
+    func testUpdatePropositionsNoNewPropositionsClearExisting() throws {
+        // setup
+        mockCache.getReturnValue = getPropositionCacheEntry()
+                
+        // test
+        mockCache.updatePropositions(nil, removing: [mockSurface])
+        
+        // verify
+        XCTAssertTrue(mockCache.removeCalled)
+        XCTAssertFalse(mockCache.setCalled)
+    }
+    
+    func testUpdatePropositionsOverwriteExistingPropositions() throws {
+        // setup
+        mockCache.getReturnValue = getPropositionCacheEntry()
+        let newSurface = mockSurface
+        let newPropositionContent = JSONFileLoader.getRulesJsonFromFile("inappPropositionV2Content")
+        let newPropositionItem = MessagingPropositionItem(itemId: "inapp4", schema: .ruleset, itemData: newPropositionContent)
+        let newProposition = MessagingProposition(uniqueId: "inapp4", scope: "inapp4", scopeDetails: ["key": "value"], items: [newPropositionItem])
+        let newProps: [Surface: [MessagingProposition]] = [
+            newSurface!: [newProposition]
+        ]
+        
+        // test
+        mockCache.updatePropositions(newProps)
+        
+        // verify
+        XCTAssertTrue(mockCache.setCalled)
+        XCTAssertEqual("propositions", mockCache.setParamKey)
+        guard let setParamCache = mockCache.setParamEntry else {
+            XCTFail("no cache parameter in 'set' call")
+            return
+        }
+        let decoder = JSONDecoder()
+        guard let decodedSetParam = try? decoder.decode([String: [MessagingProposition]].self, from: setParamCache.data) else {
+            XCTFail("failed to decode cache parameter sent during 'set' call")
+            return
+        }
+        XCTAssertEqual(1, decodedSetParam.count)
+        let paramProp = decodedSetParam["inapp2"]?.first
+        XCTAssertEqual("inapp4", paramProp?.uniqueId)
+        XCTAssertEqual("inapp4", paramProp?.scope)
+        XCTAssertEqual(1, paramProp?.items.count)
+    }
+    
+    // TODO: how to get encoding to fail (or mock it)
+    func testUpdatePropositionsBadFormatEncodeFailure() throws {
+        // setup
+        let newSurface = Surface(uri: "newSurface")
+        let newProposition = MessagingProposition(uniqueId: "", scope: "", scopeDetails: ["": ""], items: [])
+        let newProps: [Surface: [MessagingProposition]] = [
+            newSurface: [newProposition]
+        ]
+        
+        // test
+        mockCache.updatePropositions(newProps)
+        
+        // verify
+        XCTAssertFalse(mockCache.removeCalled)
+        // set should not be called but is currently because encoding doesn't fail
+        // XCTAssertFalse(mockCache.setCalled)
```

Comment:
Note: I think we can achieve this by setting up a mock override `encode` method (deriving from KeyedEncodingContainer). I'll try this out when I add tests for `MessagingProposition`